### ISSUE TITLE
agent.py questions

### ISSUE THREAD

**Issue by dan9thsense** [beyretb/AnimalAI-Olympics#45]
I am trying to figure out how to submit an entry.  In my current code, each time I want to take an action, I use this line (where I have put the desired move into the variable `action`):
`info = env.step(vector_action=action, memory=None, text_action=None)`
and then I get the observations and rewards with these lines:

```
brainInfo = info['Learner']
reward = brainInfo.rewards[0]
etc.
```

As I try to decipher the docs on submitting with docker, it looks like I need to use a class named `Agent`.  In the example, `agent.py`, the `Agent` class has a method:

`def step(self, obs, reward, done, info):`

that returns an action.  Does the evaluation software call this method and in that way get the submitted action?  Is that the place where I get the observations and reward?

Also, I am unclear on how the Agent class is supposed to interface with my code.  Is it just basically a bridge for data to pass across?  In that case, how does the simulation get started?  The dockerfile does not have a CMD at the end to get things going, so there must be some entry point elsewhere.

**Comment by beyretb** [comment]
Hi,

`agent.py` is a class that represents the basic functions your agent should implement, namely initialise (eg load any model you may have trained such as tensorflow or pytorch ones), reset (anything you need to perform at the beginning of a new episode) and take an action (based on the current observation, what action should your agent take).

We ask all participants to implement this class so we can instantiate an agent using `agent=Agent()` and call these three methods when needed. During the evaluation this is done in exactly the same way as described in the documentation, especially [here](https://github.com/beyretb/AnimalAI-Olympics/blob/master/documentation/submission.md#test-your-docker-image). Follow the submission guide step by step, have a look at [the testDocker.py file](https://github.com/beyretb/AnimalAI-Olympics/blob/master/examples/submission/test_submission/testDocker.py) to see how your agent will be used during evaluation. If your code passes the test on your local machine, it should work on the evaluation too.

>Does the evaluation software call this method and in that way get the submitted action?

yes, we give you the observation in the [Gym format](https://github.com/openai/gym), you give us the action your agent should take. 

> Is that the place where I get the observations and reward?

yes, they are in `obs` and `reward` respectively, as well as in `info` which is the same as your `brainInfo` in this case

>Also, I am unclear on how the Agent class is supposed to interface with my code. Is it just basically a bridge for data to pass across? In that case, how does the simulation get started? 

You only need to implement the `agent.py` class in your submission. We take care of all the rest, from loading an environment to running the tests.

Let us know if any of this is still unclear!

**Comment by dan9thsense** [comment]
That's helpful.  I'm using tensorflow and opencv and I know how to install those into the docker image.  However, I am wondering if there are specific version requirements to use them with the rest of the installed packages.

**Comment by beyretb** [comment]
Depends on which package we're talking about. There shouldn't be any issue with the `animalai` package, there might be conflicts with the `animalai-train` one. Only the former is used during evaluation, using the latter is up to you.

As a rule of thumb, run your docker image on [the test](https://github.com/beyretb/AnimalAI-Olympics/blob/master/documentation/submission.md#test-your-docker-image) (all packages included), if it passes, then it should work fine during evaluation.

**Comment by dan9thsense** [comment]
Thanks for the help.  Successfully submitted and scored the agent.

**Comment by shuishida** [comment]
Hello, I am looking into using AnimalAI, and wanted to see how the competition setup works. However, I can no longer find the `agent.py` or the example submission folder in general. Is there an updated documentation about the competition setup? 

Also, I was wondering if the agent can be implemented in `pytorch` instead of tensorflow 1, if I want to use the AnimalAI-train setup.

**Comment by shuishida** [comment]
Okay, I found it in the commit history and looks generic:) I am still unsure though how the training script with curriculum learning can be used to train custom agents (potentially implemented in PyTorch)
